File: lib/plotting/primitives/structural.py
"""Universal structural primitives — operate on physical shape alone.

Every primitive honors lib/plotting/theme.json (Wong/Glasbey palette,
pdf.fonttype 42, 300 dpi, provenance footer). Outputs PNG + vector PDF
byte-deterministically given the same matplotlib + freetype combo.

These five primitives cover every PhysicalShape the affordance selector
can emit as StructuralFallback:

  matrix_overview    — PhysicalShape::Numeric2D
  distribution       — PhysicalShape::Numeric1D
  categorical_summary — PhysicalShape::Categorical1D
  pairs              — PhysicalShape::TabularNumeric { columns: 2..=8 }
  scalar_card        — PhysicalShape::Scalar

Provenance kind strings match GenericPrimitive::figure_id() in
crates/core/src/plot_affordance/primitive.rs so the StructuralFallback
path resolves end-to-end without any additional wiring.

NOTE on _core.figure() / _core.savefig() discipline:
lib/plotting/core.py exposes savefig(fig, path, *, dpi, formats,
stage_id, ...) but has no figure() constructor.  Following the
established pattern of every stage module (e.g. volcano, heatmap,
bar), we call plt.subplots() directly and thread the resulting fig
through _core.savefig().  The stage_id parameter of savefig drives the
provenance footer, so we pass the primitive id string there.  Dual
PNG + PDF output is achieved by passing formats=["png", "pdf"] and
the PNG path as the primary; savefig writes the PDF alongside it with
the same stem (path.with_suffix(".pdf")).
"""

from __future__ import annotations

from collections import Counter
from pathlib import Path
from typing import Any, Sequence

import matplotlib.pyplot as plt
import numpy as np

# Import the module-level object so we get the live THEME and all
# helpers without binding to a specific attribute at import time.
# runtime.plotting is available when executed inside an emitted package;
# fall back to lib.plotting when run from the repo root (e.g. pytest).
try:
    from runtime.plotting import core as _core
except ImportError:
    from lib.plotting import core as _core  # type: ignore[no-redef]


def matrix_overview(
    matrix: np.ndarray,
    *,
    png_path: Path,
    pdf_path: Path,
    title: str,
    theme_path: str,
) -> None:
    """Heatmap of any 2D numeric array.

    Rasterizes the imshow layer when matrix.size > 50_000 to keep the
    PDF file size manageable while preserving a vector frame and labels.
    The colorbar, title, and axis labels remain vector in both cases.

    Args:
        matrix:     2D numpy array of numeric values.
        png_path:   Destination path for the PNG output (300 dpi).
        pdf_path:   Destination path for the PDF output (vector).
        title:      Figure title shown above the heatmap.
        theme_path: Theme identifier forwarded for future preset
                    resolution; ignored by the current implementation
                    which reads the module-level THEME directly.
    """
    if np.asarray(matrix).ndim != 2:
        raise ValueError(
            f"matrix_overview expects a 2D input array; got shape {np.asarray(matrix).shape}"
        )
    mat = np.asarray(matrix, dtype=float)
    rasterize = mat.size > 50_000

    fig, ax = plt.subplots(figsize=(8.0, 6.0))
    im = ax.imshow(mat, aspect="auto", rasterized=rasterize)
    fig.colorbar(im, ax=ax)
    ax.set_title(title)
    ax.set_xlabel("column")
    ax.set_ylabel("row")

    # Savefig writes PNG + PDF via formats=["png","pdf"]. The primary
    # path carries the PNG suffix; the companion PDF is written alongside
    # With the same stem. stage_id drives the provenance footer.
    _core.savefig(
        fig,
        Path(png_path),
        stage_id="__structural_matrix_overview",
        formats=["png", "pdf"],
    )


def distribution(
    values: np.ndarray,
    *,
    png_path: Path,
    pdf_path: Path,
    title: str,
    theme_path: str,
    bins: int = 50,
) -> None:
    """Histogram of any 1D numeric array with an optional KDE overlay.

    The KDE overlay is added only when n >= 25 — below that threshold
    the kernel estimate is too unstable to be informative.

    Args:
        values:     1D (or ravelled) numeric array.
        png_path:   Destination path for the PNG output.
        pdf_path:   Destination path for the PDF output.
        title:      Figure title.
        theme_path: Theme identifier (forwarded; not used directly).
        bins:       Number of histogram bins (default 50).
    """
    arr = np.asarray(values, dtype=float).ravel()
    if arr.size == 0:
        raise ValueError("distribution requires a non-empty input array")

    fig, ax = plt.subplots(figsize=(8.0, 5.5))
    ax.hist(arr, bins=bins, density=True, alpha=0.6)

    if arr.size >= 25:
        from scipy.stats import gaussian_kde  # lazy import — optional dep path

        kde = gaussian_kde(arr)
        xs = np.linspace(float(arr.min()), float(arr.max()), 200)
        ax.plot(xs, kde(xs))

    ax.set_title(title)
    ax.set_xlabel("value")
    ax.set_ylabel("density")

    _core.savefig(
        fig,
        Path(png_path),
        stage_id="__structural_distribution",
        formats=["png", "pdf"],
    )


def categorical_summary(
    labels: Sequence[Any],
    *,
    png_path: Path,
    pdf_path: Path,
    title: str,
    theme_path: str,
) -> None:
    """Bar plot showing per-category counts of a 1D categorical array.

    Categories are sorted by descending count; ties are broken by
    ascending string representation so the order is fully deterministic
    regardless of insertion order.

    Args:
        labels:     Sequence of hashable category labels.
        png_path:   Destination path for the PNG output.
        pdf_path:   Destination path for the PDF output.
        title:      Figure title.
        theme_path: Theme identifier (forwarded; not used directly).
    """
    counts = Counter(labels)
    if not counts:
        raise ValueError("categorical_summary requires at least one label")

    # Sort: descending count, then ascending string label as tie-breaker.
    sorted_items = sorted(counts.items(), key=lambda kv: (-kv[1], str(kv[0])))
    keys = [str(kv[0]) for kv in sorted_items]
    vals = [kv[1] for kv in sorted_items]

    n_cats = len(keys)
    fig_width = max(6.0, min(16.0, n_cats * 0.6 + 2.0))
    fig, ax = plt.subplots(figsize=(fig_width, 5.5))

    palette = _core.categorical_palette(n_cats, name="categorical_summary")
    ax.bar(range(n_cats), vals, color=palette)
    ax.set_xticks(range(n_cats))
    ax.set_xticklabels(keys, rotation=45, ha="right", fontsize=7)
    ax.set_title(title)
    ax.set_ylabel("count")

    _core.savefig(
        fig,
        Path(png_path),
        stage_id="__structural_categorical_summary",
        formats=["png", "pdf"],
    )


def pairs(
    table: np.ndarray,
    column_names: Sequence[str],
    *,
    png_path: Path,
    pdf_path: Path,
    title: str,
    theme_path: str,
) -> None:
    """Small-multiples scatter matrix for 2D tabular numeric input.

    Diagonal cells show per-column histograms; off-diagonal cells show
    pairwise scatter plots.  Scatter points are rasterized to keep PDF
    file size bounded.

    The column count is validated to be <= 8.  At higher cardinality
    the individual panels become too small to be legible and the caller
    should aggregate columns before plotting.

    Args:
        table:        2D array of shape (n_rows, n_cols) with n_cols <= 8.
        column_names: Sequence of n_cols human-readable column labels.
        png_path:     Destination path for the PNG output.
        pdf_path:     Destination path for the PDF output.
        title:        Figure suptitle.
        theme_path:   Theme identifier (forwarded; not used directly).
    """
    arr = np.asarray(table, dtype=float)
    if arr.ndim != 2:
        raise ValueError(
            f"pairs expects a 2D input array; got ndim={arr.ndim}"
        )
    n_cols = arr.shape[1]
    if n_cols > 8:
        raise ValueError(
            f"pairs accepts at most 8 columns; got {n_cols}. "
            "Reduce the number of columns before calling pairs()."
        )
    if len(column_names) != n_cols:
        raise ValueError(
            f"column_names length ({len(column_names)}) must match "
            f"table column count ({n_cols})"
        )

    cell_size = 2.0
    fig_size = cell_size * n_cols
    fig, axes = plt.subplots(n_cols, n_cols, figsize=(fig_size, fig_size))

    # Normalise axes to always be a 2D array for uniform indexing.
    if n_cols == 1:
        axes = np.array([[axes]])
    else:
        axes = np.asarray(axes)

    scatter_color = _core.categorical_palette(1, name="pairs")[0]

    for i in range(n_cols):
        for j in range(n_cols):
            ax = axes[i, j]
            if i == j:
                ax.hist(arr[:, i], bins=30, color=scatter_color, alpha=0.7)
            else:
                ax.scatter(
                    arr[:, j],
                    arr[:, i],
                    s=4,
                    alpha=0.5,
                    color=scatter_color,
                    rasterized=True,
                )
            # Only show tick labels on the outer edges to reduce clutter.
            if i < n_cols - 1:
                ax.set_xticklabels([])
            else:
                ax.set_xlabel(column_names[j], fontsize=7)
            if j > 0:
                ax.set_yticklabels([])
            else:
                ax.set_ylabel(column_names[i], fontsize=7)

    fig.suptitle(title, y=1.01)
    fig.tight_layout()

    _core.savefig(
        fig,
        Path(png_path),
        stage_id="__structural_pairs",
        formats=["png", "pdf"],
    )


def scalar_card(
    value: float,
    label: str,
    *,
    png_path: Path,
    pdf_path: Path,
    title: str,
    theme_path: str,
) -> None:
    """Plain-text display card for a single scalar metric.

    Renders the numeric value in large type with a descriptive label
    below it, suitable for surfacing aggregate statistics (e.g. AUROC,
    R-squared, p-value) in the result-review pane.

    Args:
        value:      The scalar value to display.
        label:      Short descriptive label for the value.
        png_path:   Destination path for the PNG output.
        pdf_path:   Destination path for the PDF output.
        title:      Figure title shown above the card.
        theme_path: Theme identifier (forwarded; not used directly).
    """
    fig, ax = plt.subplots(figsize=(4.0, 2.5))
    ax.axis("off")
    ax.text(
        0.5,
        0.60,
        f"{value:.4g}",
        ha="center",
        va="center",
        fontsize=36,
        transform=ax.transAxes,
    )
    ax.text(
        0.5,
        0.25,
        label,
        ha="center",
        va="center",
        fontsize=14,
        transform=ax.transAxes,
    )
    ax.set_title(title)

    _core.savefig(
        fig,
        Path(png_path),
        stage_id="__structural_scalar_card",
        formats=["png", "pdf"],
    )

File: crates/ecaa-conformance/tests/fixtures/minimal-package-real/runtime/plotting_r/primitives/structural.R
# Universal structural primitives — R parity to
# lib/plotting/primitives/structural.py.
#
# Every primitive sources lib/plotting_r/core.R for theme, palette,
# and swfc_savefig so visual style is byte-stable given the same
# pinned ggplot2 + Cairo + ragg versions.
#
# Provenance kind strings match GenericPrimitive::figure_id() in
# crates/core/src/plot_affordance/primitive.rs:
#   __structural_matrix_overview
#   __structural_distribution
#   __structural_categorical_summary
#   __structural_pairs
#   __structural_scalar_card
#
# Each function signature mirrors the Python counterpart:
#   primitive_name(data, ..., png_path, pdf_path, title, theme_path)
#
# NOTE: GGally and gridExtra are not declared as deps in DESCRIPTION.
# The pairs() primitive therefore uses a base ggplot2 facet approach
# via a long-format melt + facet_grid, which requires no additional
# packages.
#
# swfc_savefig() writes both PNG and PDF when THEME$output$formats
# contains both entries (the default).  The primary path passed is the
# PNG path; the companion PDF is written alongside with the same stem
# (matching the Python discipline in savefig(formats=["png","pdf"])).
#
# Guard: source core.R before this file.

if (!exists("swfc_savefig")) {
  stop(paste0(
    "source runtime/plotting_r/core.R before sourcing this file.\n",
    "e.g.  source('lib/plotting_r/core.R')\n",
    "      source('lib/plotting_r/primitives/structural.R')"
  ))
}


# ---------------------------------------------------------------------------
# matrix_overview — PhysicalShape::Numeric2D
# ---------------------------------------------------------------------------

#' Heatmap of any 2D numeric matrix.
#'
#' Rasterizes the tile layer when matrix size > 50 000 cells (controlled by
#' THEME$output$rasterize_threshold_n) to keep PDF file size bounded.
#'
#' @param matrix  Numeric matrix (rows x cols).
#' @param png_path  Character path for the PNG output.
#' @param pdf_path  Character path for the PDF output (written alongside PNG).
#' @param title     Figure title.
#' @param theme_path  Theme identifier string (forwarded; THEME is read
#'                    from the live module-level object loaded by core.R).
structural_matrix_overview <- function(
    matrix,
    png_path,
    pdf_path,
    title      = "",
    theme_path = "theme.json"
) {
  if (!is.matrix(matrix) && !is.array(matrix)) {
    matrix <- as.matrix(matrix)
  }
  if (length(dim(matrix)) != 2) {
    stop(sprintf(
      "structural_matrix_overview expects a 2D matrix; got %d dimensions",
      length(dim(matrix))
    ))
  }
  n_rows <- nrow(matrix)
  n_cols <- ncol(matrix)
  thresh <- THEME$output$rasterize_threshold_n %||% 50000L

  # Convert matrix to long-format data frame for ggplot2 geom_tile.
  df <- data.frame(
    row   = rep(seq_len(n_rows), times = n_cols),
    col   = rep(seq_len(n_cols), each  = n_rows),
    value = as.vector(matrix),
    stringsAsFactors = FALSE
  )

  rasterize <- (n_rows * n_cols) > thresh

  p <- ggplot2::ggplot(df, ggplot2::aes(x = .data$col, y = .data$row, fill = .data$value)) +
    ggplot2::geom_tile(raster = rasterize) +
    ggplot2::scale_fill_viridis_c(name = "value") +
    ggplot2::scale_y_reverse() +
    ggplot2::labs(title = title, x = "column", y = "row") +
    ggplot2::theme(aspect.ratio = n_rows / n_cols)

  p <- .swfc_attach_footer(p, "__structural_matrix_overview")
  swfc_savefig(p, png_path, stage_id = "__structural_matrix_overview")
  invisible(NULL)
}


# ---------------------------------------------------------------------------
# distribution — PhysicalShape::Numeric1D
# ---------------------------------------------------------------------------

#' Histogram of any 1D numeric vector with optional KDE overlay.
#'
#' The KDE overlay is added only when length(values) >= 25.
#'
#' @param values    Numeric vector.
#' @param png_path  Character path for the PNG output.
#' @param pdf_path  Character path for the PDF output.
#' @param title     Figure title.
#' @param theme_path  Theme identifier string (forwarded; not used directly).
#' @param bins      Number of histogram bins (default 50).
structural_distribution <- function(
    values,
    png_path,
    pdf_path,
    title      = "",
    theme_path = "theme.json",
    bins       = 50L
) {
  arr <- as.numeric(values)
  arr <- arr[is.finite(arr)]
  if (length(arr) == 0) {
    stop("structural_distribution requires a non-empty input vector")
  }

  df <- data.frame(value = arr, stringsAsFactors = FALSE)

  p <- ggplot2::ggplot(df, ggplot2::aes(x = .data$value)) +
    ggplot2::geom_histogram(
      ggplot2::aes(y = ggplot2::after_stat(density)),
      bins  = bins,
      fill  = .WONG_PALETTE[6],
      color = "#333333",
      linewidth = 0.2,
      alpha = 0.6
    ) +
    ggplot2::labs(title = title, x = "value", y = "density")

  if (length(arr) >= 25L) {
    p <- p + ggplot2::geom_density(
      color     = .WONG_PALETTE[3],
      linewidth = 0.7
    )
  }

  p <- .swfc_attach_footer(p, "__structural_distribution")
  swfc_savefig(p, png_path, stage_id = "__structural_distribution")
  invisible(NULL)
}


# ---------------------------------------------------------------------------
# categorical_summary — PhysicalShape::Categorical1D
# ---------------------------------------------------------------------------

#' Bar plot of per-category counts for a 1D categorical vector.
#'
#' Categories are sorted by descending count; ties broken by ascending
#' string representation (deterministic regardless of insertion order).
#'
#' @param labels    Character (or factor / integer) vector of category labels.
#' @param png_path  Character path for the PNG output.
#' @param pdf_path  Character path for the PDF output.
#' @param title     Figure title.
#' @param theme_path  Theme identifier string (forwarded; not used directly).
structural_categorical_summary <- function(
    labels,
    png_path,
    pdf_path,
    title      = "",
    theme_path = "theme.json"
) {
  labels <- as.character(labels)
  if (length(labels) == 0) {
    stop("structural_categorical_summary requires at least one label")
  }

  counts_tbl <- sort(table(labels), decreasing = TRUE)
  # Deterministic tie-break: sort tied groups by ascending label string.
  freq <- as.integer(counts_tbl)
  nms  <- names(counts_tbl)
  ord  <- order(-freq, nms)
  freq <- freq[ord]
  nms  <- nms[ord]

  df <- data.frame(
    category = factor(nms, levels = nms),
    count    = freq,
    stringsAsFactors = FALSE
  )

  n_cats <- length(nms)
  pal    <- swfc_palette(n_cats, name = "categorical_summary")

  fig_width <- max(6.0, min(16.0, n_cats * 0.6 + 2.0))

  p <- ggplot2::ggplot(df, ggplot2::aes(x = .data$category, y = .data$count,
                                        fill = .data$category)) +
    ggplot2::geom_col(width = 0.75, color = "#333333", linewidth = 0.2) +
    ggplot2::scale_fill_manual(values = stats::setNames(pal, nms), guide = "none") +
    ggplot2::labs(title = title, x = "", y = "count") +
    ggplot2::theme(axis.text.x = ggplot2::element_text(angle = 45, hjust = 1, size = 7))

  p <- .swfc_attach_footer(p, "__structural_categorical_summary")
  swfc_savefig(p, png_path,
               stage_id = "__structural_categorical_summary",
               width_in  = fig_width,
               height_in = 5.5)
  invisible(NULL)
}


# ---------------------------------------------------------------------------
# pairs — PhysicalShape::TabularNumeric { columns: 2..=8 }
# ---------------------------------------------------------------------------

#' Small-multiples scatter matrix for 2D tabular numeric input (<=8 columns).
#'
#' Diagonal cells show per-column density plots; off-diagonal cells show
#' pairwise scatter plots.  Implemented as a ggplot2 facet_grid over
#' a melted long-format data frame — no GGally or gridExtra dependency.
#'
#' @param table        Numeric matrix or data.frame of shape (n_rows, n_cols).
#'                     n_cols must be between 2 and 8 inclusive.
#' @param column_names Character vector of length n_cols.
#' @param png_path     Character path for the PNG output.
#' @param pdf_path     Character path for the PDF output.
#' @param title        Figure suptitle.
#' @param theme_path   Theme identifier string (forwarded; not used directly).
structural_pairs <- function(
    table,
    column_names,
    png_path,
    pdf_path,
    title      = "",
    theme_path = "theme.json"
) {
  mat <- as.matrix(table)
  if (length(dim(mat)) != 2) {
    stop("structural_pairs expects a 2D matrix or data.frame")
  }
  n_cols <- ncol(mat)
  if (n_cols > 8L) {
    stop(sprintf(
      "structural_pairs accepts at most 8 columns; got %d. ",
      n_cols
    ))
  }
  if (length(column_names) != n_cols) {
    stop(sprintf(
      "column_names length (%d) must match table column count (%d)",
      length(column_names), n_cols
    ))
  }

  colnames(mat) <- column_names
  scatter_color <- .WONG_PALETTE[6]

  # Build each panel as an individual ggplot object and arrange via
  # a grid.  We use base graphics grid.newpage / pushViewport so we
  # need only the standard grid package (already a dep via ggplot2).
  cell_plots <- vector("list", n_cols * n_cols)
  for (i in seq_len(n_cols)) {
    for (j in seq_len(n_cols)) {
      idx <- (i - 1) * n_cols + j
      xi <- mat[, j]
      yi <- mat[, i]
      if (i == j) {
        df_diag <- data.frame(v = xi, stringsAsFactors = FALSE)
        cell_plots[[idx]] <- ggplot2::ggplot(df_diag, ggplot2::aes(x = .data$v)) +
          ggplot2::geom_histogram(bins = 30L, fill = scatter_color,
                                  color = "#333333", linewidth = 0.2, alpha = 0.7) +
          ggplot2::labs(x = if (i == n_cols) column_names[j] else NULL,
                        y = if (j == 1)    column_names[i] else NULL) +
          ggplot2::theme(
            axis.title.x = ggplot2::element_text(size = 7),
            axis.title.y = ggplot2::element_text(size = 7),
            plot.margin  = grid::unit(c(1, 1, 1, 1), "pt")
          )
      } else {
        df_sc <- data.frame(x = xi, y = yi, stringsAsFactors = FALSE)
        cell_plots[[idx]] <- ggplot2::ggplot(df_sc, ggplot2::aes(x = .data$x, y = .data$y)) +
          ggplot2::geom_point(size = 0.8, alpha = 0.5, color = scatter_color, stroke = 0) +
          ggplot2::labs(x = if (i == n_cols) column_names[j] else NULL,
                        y = if (j == 1)    column_names[i] else NULL) +
          ggplot2::theme(
            axis.title.x = ggplot2::element_text(size = 7),
            axis.title.y = ggplot2::element_text(size = 7),
            plot.margin  = grid::unit(c(1, 1, 1, 1), "pt")
          )
      }
    }
  }

  # Combine using patchwork if available; otherwise fall back to grid.
  cell_size <- 2.0
  fig_size  <- cell_size * n_cols

  if (requireNamespace("patchwork", quietly = TRUE)) {
    combined <- Reduce(`+`, cell_plots) +
      patchwork::plot_layout(ncol = n_cols) +
      patchwork::plot_annotation(title = title)
    swfc_savefig(combined, png_path,
                 stage_id  = "__structural_pairs",
                 width_in  = fig_size,
                 height_in = fig_size)
  } else {
    # Fallback: write each cell to a temporary file and compose with grid.
    png_out <- as.character(png_path)
    pdf_out <- sub("\\.[^.]+$", ".pdf", png_out)
    dpi     <- THEME$output$png_dpi %||% 300L

    .swfc_write_pairs_grid(cell_plots, n_cols, title,
                           png_out, pdf_out, fig_size, dpi)
  }
  invisible(NULL)
}


#' Internal grid-based layout used when patchwork is unavailable.
#' @keywords internal
.swfc_write_pairs_grid <- function(cell_plots, n_cols, title,
                                   png_out, pdf_out, fig_size, dpi) {
  .write_one <- function(out_path, device_fn) {
    device_fn(out_path, width = fig_size, height = fig_size, units = "in", res = dpi)
    on.exit(grDevices::dev.off(), add = TRUE)
    grid::grid.newpage()
    grid::pushViewport(
      grid::viewport(layout = grid::grid.layout(n_cols, n_cols))
    )
    for (idx in seq_along(cell_plots)) {
      i <- ((idx - 1) %/% n_cols) + 1L
      j <- ((idx - 1) %%  n_cols) + 1L
      grid::pushViewport(grid::viewport(
        layout.pos.row = i, layout.pos.col = j
      ))
      print(cell_plots[[idx]], newpage = FALSE)
      grid::popViewport()
    }
    grid::popViewport()
    if (nzchar(title)) {
      grid::grid.text(title, x = 0.5, y = 0.99,
                      just = c("centre", "top"),
                      gp = grid::gpar(fontsize = 9))
    }
  }

  .write_one(png_out, function(p, ...) ragg::agg_png(p, ..., background = "white"))
  .write_one(pdf_out, function(p, width, height, ...) {
    grDevices::cairo_pdf(p, width = width, height = height, onefile = FALSE,
                         family = THEME$fonts$stack[[1]] %||% "sans")
  })
}


# ---------------------------------------------------------------------------
# scalar_card — PhysicalShape::Scalar
# ---------------------------------------------------------------------------

#' Plain-text display card for a single scalar metric.
#'
#' Renders the numeric value in large type with a descriptive label below,
#' suitable for surfacing aggregate statistics (AUROC, R-squared, p-value).
#'
#' @param value     Scalar numeric value.
#' @param label     Short descriptive label.
#' @param png_path  Character path for the PNG output.
#' @param pdf_path  Character path for the PDF output.
#' @param title     Figure title.
#' @param theme_path  Theme identifier string (forwarded; not used directly).
structural_scalar_card <- function(
    value,
    label,
    png_path,
    pdf_path,
    title      = "",
    theme_path = "theme.json"
) {
  value_str <- formatC(as.numeric(value), digits = 4, format = "g")

  df_val   <- data.frame(x = 0.5, y = 0.60, label = value_str,
                         stringsAsFactors = FALSE)
  df_label <- data.frame(x = 0.5, y = 0.25, label = as.character(label),
                         stringsAsFactors = FALSE)

  p <- ggplot2::ggplot() +
    ggplot2::xlim(0, 1) + ggplot2::ylim(0, 1) +
    ggplot2::geom_text(
      data = df_val,
      ggplot2::aes(x = .data$x, y = .data$y, label = .data$label),
      size  = 36 / .pt,
      color = "#222222"
    ) +
    ggplot2::geom_text(
      data = df_label,
      ggplot2::aes(x = .data$x, y = .data$y, label = .data$label),
      size  = 14 / .pt,
      color = "#444444"
    ) +
    ggplot2::labs(title = title) +
    ggplot2::theme_void() +
    ggplot2::theme(
      plot.title   = ggplot2::element_text(size = THEME$fonts$title_pt %||% 9,
                                           hjust = 0.5),
      plot.margin  = grid::unit(c(8, 8, 24, 8), "pt")
    )

  p <- .swfc_attach_footer(p, "__structural_scalar_card")
  swfc_savefig(p, png_path,
               stage_id  = "__structural_scalar_card",
               width_in  = 4.0,
               height_in = 2.5)
  invisible(NULL)
}
